org/mozilla/jss/util/NativeEnclosure.java

```diff
@@ -0,0 +1,46 @@
+package org.mozilla.jss.util;
+
+/**
+ * Classes extending NativeEnclsoure wrap a single NativeProxy instance,
+ * allowing it to be accessed from the JNI layer but be allocated and scoped
+ * from the Java layer.
+ *
+ * This allows us to data parameters in Java and call open(...) when all
+ * necessary variables have been created. At the end of this object's use,
+ * close must be called to free allocated memory.
+ *
+ * Because this class implements AutoCloseable, it is suggested to add
+ * constructors to derived classes which call open; this'll allow a single
+ * try-with-resources block to scope the lifetime of this object:
+ *
+ * <pre>
+ * try (NEC obj = new NEC(...)) {
+ *      // ... do something with obj ...
+ * }
+ * </pre>
+ */
+public abstract class NativeEnclosure implements AutoCloseable {
+    public NativeProxy ptr;
+    public long ptr_size;
+
+    public final void open() throws Exception {
+        if (ptr == null) {
```

Comment:
How does `ptr` get initialized? Should we make it more explicit like this?
```
if (ptr == null) {
    ptr = acquireNativeResources();
}

if (ptr != null) {
    releaseNativeResources();
    ptr.close();
    ptr = null;
}
```

Just thinking, does it make sense to move `acquireNativeResources()` and `releaseNativeResources()` into `NativeProxy`?

Reply:
A `NativeProxy` is a way of allocating, tracking, and giving control of that pointer to Java. The constructor for `NativeProxy` takes a pointer, tracks it, and provides a `close(...)` method to free it. That means--on construction of the object--this pointer needs to be non-null and provided. In order to release resources, `NativeProxy` has an abstract `releaseNativeResources()` method which calls `free(...)` or whatever else is correct in this situation. Derived classes implement that method. Examples of `NativeProxy` include [`GlobalRefProxy`](https://github.com/dogtagpki/jss/blob/master/org/mozilla/jss/util/GlobalRefProxy.java), [`TokenProxy`](https://github.com/dogtagpki/jss/blob/master/org/mozilla/jss/pkcs11/TokenProxy.java), and all of the PK11 key types.

Typically, the proxy is allocated by something else -- `CryptoManager` creates a number of `PK11Token` instances based on the local modules and stores them in the `TokenSupplier` singleton.

---

A `StaticVoidRef` inverts the first property of `NativeProxy`s: that they implement free. In particular, the `releaseNativeResources()` call is a no-op. This is useful when you have C code like:

```
{
    InsaneType obj;
    staticvoidref = wrap(&obj);
    call_some_jni(staticvoidref);
}
```

In this case, since InsaneType is scoped and doesn't need a free call (with the assumption that the JNI call doesn't store a copy of the reference, it just needs it for whatever reason), this is safe and fine.

Another example of its usage would be in the JSSTrustManager code with the legacy SSLSocket: you allocate `CERTCertificate`s in the NSS certificate handler layer, but you need to pass them to the JSSTrustManager which doesn't store them and instead uses them only for the lifetime of the JNI call. You'd ideally keep the scope of the `CERTCertificate` in the NSS certificate handler, and never free it in the JSSTrustManager because you'd have no way of knowing if it was free'd. Using a `StaticVoidRef` could work there (other than the slight fact issue that it doesn't type check but that's a different issue). 

Lastly, because it isn't something that needs to be free'd later, it can't be tracked. Sometimes a void pointer is useful in a return parameter in C code. In that case, you'd allocate a `StaticVoidRef` object with a `NULL` pointer, pass it to JNI, but since `mPointer == NULL`, tracking could fail and might involve a collision. That's not good! So we extend `NativeProxy` to disable tracking temporarily.

--- 

Lastly we get to `NativeEnclosure`: a `NativeProxy` is a way for C code to give up a pointer it created and tell Java that it gets to deal with the scope of it. However, _something_ still needs to tell C to create said pointer. That's where `NativeEnclosure` comes in: it is a contract for Java to tell C, hey, do your thing. It "encloses" the scope of the pointer: something calls `open(...)` to create `ptr` and then something calls `close(...)` to free it, but tell Java when you do either.

We use `NativeEnclosure` in conjunction with `StaticVoidRef` to create the following situation:

 - Data parameters are Java classes with constructors and setters. (this is while `ptr == null`)
 - Data parameters need to get wrapped into a C pointer "just in time" for the `derive_key` call. (this is `open(...)`)
 - Occasionally, before the C pointer is free'd, we need to exfiltrate data; this is a necessity with additional derived keys. That's why `this.releaseNativeResources(...)` is called called head of `ptr.close()`. 
     - However, since we're already getting the value of `ptr`, we might as well clean it up and free it while were here. That's why we use `StaticVoidRef` -- to prevent a double free.
 - We also can't double-free. If ptr is non-null and is of type other than `StaticVoidRef`, we'd want to track that. Hence `close(...)` only works if `ptr != null`.
 - We also can't lose memory. If `ptr` is null, you'd first have to call `close(...)` before `open(...)` works again.

Hope that makes some sense?

Reply:
Concretely: 

> How does `ptr` get initialized? 

`ptr` gets initialized from `acquireNativeResources(...)` which is called only by open. `acquireNativeResources(...)` is an implementation detail of the derived enclosing type.

`ptr` needs to be public though, so that other (JNI) methods can access its value. This happens with e.g., `KBKDFCounterParam`, where it has an array of `KBKDFDataParameters` which it calls `open()` on and in turn calls `close()` on when it gets free'd.

Reply:
I guess for an analogy:

`NativeEnclosure` is like curly braces...

```c
{ // Begin enclosure
int something;
void *ptr; 

... do some things ... 

// Now you have a choice: what do you put inside the enclosure?
// Do you put a reference to something statically allocated?
ptr = &something;

... do some other things ... 

// or do you dynamically allocate memory?
ptr = malloc(sizeof(int));

... do some more things ... 

// ... but if you did, you better call free:
free(ptr);

... do something else entirely ... 

} // End Enclosure 
```

`acquireNativeResouces(...)` behaves like an assignment or initialization function (e.g., `CMAC_Init(...)`). `open(...)` protects you from calling `acquireNativeResources(...)` multiple times without calling `close(...)`. `releaseNativeResources(...)` acts like a `free(...)` call / custom deconstruction handler (e.g., `CMAC_Destroy(...)`). `close(...)` protects you from calling `releaseNativeResources(...)` and `free(...)` multiple times.

Reply:
Thanks for the detailed explanation! I can't say I understand everything :) but I'm just wondering, if `acquireNativeResources()` is responsible to initialize `ptr`, should we change it to return the value of the `ptr` like this?
```
ptr = acquireNativeResources();
```
This will ensure the the `ptr` will be set properly, avoiding bugs in the subclass.

Reply:
There's two fields though:

```Java
public abstract class NativeEnclosure implements AutoCloseable {
    public NativeProxy ptr;
    public long ptr_size;
...
}
```

---

(aside)

In Java, you don't particularly care about the size of your object. When size is important (e.g., if its an List or an Array or something), you have attributes to find it. Also, except for a few primitive types (`int`, `int[]`, `boolean`, &c), Everything is an object.

In C, you almost always need to know the size of your pointer and what type of pointer it is, especially for memory management. The two exceptions are NULL-terminated C strings and calling `free(...)`. Regular arrays almost always have a size attached to them, pointers to structs might have a size attached to make sure that you weren't passed garbage, and a myriad of other uses.

Also, there's two "types" of pointers: a reference to the address space of a variable, typically used to return a value out of a function, and a reference to a contiguous address space of `n` elements, i.e., an array.

`NativeProxy` is just a proxy for the pointer and so it lacks size information.

`NativeEnclosure` wraps a pointer (via `NativeProxy`) _and_ includes the size information for that pointer, so you can determine if you got a 1-sized array or an n-sized array, _assuming_ you know the type of the pointer. :)

---

Anyhow, all this to say, we could return `ptr`, but we can't return both `ptr` and `ptr_size`. I can't do something like I would in C:

```
<error type> acquireNativeResources(&ptr, &ptr_size);
```

You could do something similar to the above if you changed `ptr_size` to a `Long`, but they're instance variables anyways, so I'm not sure that'd be very helpful...

I've added some comments to clarify this some more hopefully.